### PR TITLE
add dynamic debounce support

### PR DESCRIPTION
**Pull request by ramaraochavali**
Implements https://github.com/istio/istio/issues/24021 proposal with config flag defaulted to false The default behaviour is to retain the current debounce logic.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/features/pilot.go [hunk 1]
```diff
@@ -82,6 +82,12 @@ var (
 			" EDS pushes may be delayed, but there will be fewer pushes. By default this is enabled",
 	)
 
+	EnableDynamicDebounce = env.RegisterBoolVar(
+		"PILOT_ENABLE_DYNAMIC_DEBOUNCE",
+		true,
```
Comment (sdake) [inline pilot/pkg/features/pilot.go:87]: please change to default false before merging.

File: pilot/pkg/features/pilot.go [hunk 2]
```diff
@@ -70,7 +70,7 @@ var (
 
 	DebounceMax = env.RegisterDurationVar(
 		"PILOT_DEBOUNCE_MAX",
-		10*time.Second,
+		5*time.Second,
```
Comment (sdake) [inline pilot/pkg/features/pilot.go:73]: if you think the debounce default max requires a change, could you break out this one liner as a separate PR?

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 3]
```diff
@@ -284,6 +286,10 @@ func debounce(ch chan *model.PushRequest, stopCh <-chan struct{}, pushFn func(re
 
 	pushCounter := 0
 	debouncedEvents := 0
+	// back-off for deboumce. For fixed debouncer, it will be constant.
+	// When dynamic debouncer is enabled, this will be incremental
```
Comment (sdake) [inline pilot/pkg/proxy/envoy/v2/discovery.go:290]: instead of incremental - why not add some randomness?

**Comment by ramaraochavali** [comment]
@sdake I am trying to achieve what is documented here https://github.com/istio/istio/issues/24021#issue-622751008. I enabled the new behaviour by default to see how tests behave.  I am seeing some tests are very sensitive to current debounce behaviour.  Once I figure that part out, I will switch the default back so that we can verify perf and decice.

Regarding PR title - I had an earlier attempt of doing this and this PR is simpler than that :-) . I agree it is misleading. I will correct it once I make all tests pass with the new behaviour.

File: pilot/pkg/features/pilot.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/features/pilot.go:73]: No. I do not intend to change this. I am experimenting with few values for some of the sensitive tests

**Comment by ramaraochavali** [comment]
@howardjohn This is ready for review. All tests are passing with both modes. PTAL when you get chance. It would be good if we have a perf test that can validate this behaviour

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 4]
```diff
@@ -300,23 +307,34 @@ func debounce(ch chan *model.PushRequest, stopCh <-chan struct{}, pushFn func(re
 	}
 
 	pushWorker := func() {
-		eventDelay := time.Since(startDebounce)
+		debounceDelay := time.Since(startDebounce)
 		quietTime := time.Since(lastConfigUpdateTime)
-		// it has been too long or quiet enough
-		if eventDelay >= debounceMax || quietTime >= debounceAfter {
+		// it has been too long or quiet enough - push the current requests.
+		if debounceDelay >= debounceMax || quietTime >= debounceBackoff {
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/discovery.go:313]: Not sure this is right. If we do exponential back off and let’s say we are currently at 8s, what this will mean is that after the last apply from a large yaml file, there will be 8s of delay to push configs.

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 5]
```diff
@@ -300,23 +307,34 @@ func debounce(ch chan *model.PushRequest, stopCh <-chan struct{}, pushFn func(re
...
 			}
 		} else {
-			timeChan = time.After(debounceAfter - quietTime)
+			if enableDynamicDebounce {
+				// If dynamic debounce is enabled, backoff slowly like 100ms, 200ms, 400ms etc,
+				// but do not go beyond max debounce.
+				debounceBackoff *= 2
+				if debounceBackoff >= debounceMax {
+					debounceBackoff = debounceMax
+				}
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/discovery.go:333]: What’s the max here?

Comment (ramaraochavali) [reply pilot/pkg/proxy/envoy/v2/discovery.go:333]: 10s - default max

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 4 above)
Comment (ramaraochavali) [reply pilot/pkg/proxy/envoy/v2/discovery.go:313]: Do you think we should stick with the old approach while determining the quiet time i.e. if there are no events in the `debounceAfter` (100ms) - we just push the config? I was debating about that  - If that is better, I can change it. or any other better ideas?

**Comment by howardjohn** [comment]
So the case I was thinking of was `kubectl apply -f some-large-file`. I have seen often this results in over 100ms between configs. For example here lets assume 200ms between config changes

Today, if you apply this it will send every config, nothing debounce
With this PR, it will still send every config, because we never actually get multiple configs applied within the same 100ms. If we got lucky and 2 configs apply quickly we may trigger this, but if its exactly 200ms then we never will. So essentially this is backing off within a "debounce cycle" (I just made this up - not sure if there is a better word)

I was thinking more along the lines of backing off between cycles. So something along the lines of first config applies, we wait for 100ms. Once 100ms is up, we trigger a push. Then another config change comes in. At this point, we check if its been X time since the last config change (not sure what X should be). If greater than X time, we half the debounce time (or maybe reset). If its been less than X seconds, we 2x the debounce time. What this results in is:

100ms - push
200ms - push
400ms - push
800ms - push

Whereas with your PR we will just do a push at 800ms exclusively.

I am not sure which of these is best - I can see both being useful which is why I was hoping we would have an intern implement https://github.com/istio/istio/issues/20123 to test this but I don't think that is happening now.

**Comment by rshriram** [comment]
This is an interesting problem. Let’s solve it in an abstract sense and then plug the solution(s).

If we take the kubectl scenario, best case imo is ONE push after the entire yaml file has been pushed. This ensures consistent behavior (eg push happens only after both virtual service and destination rule are in). Worst case is what we have (push for each config).

What if we increase de-bounce period to 250ms?
Or build a simple predictor that looks back at last N events to predict when next event would happen ?  If prediction fails or estimated arrival time of next event is too far off, then we push. The current system we have is essentially a predictor that looks back by one event.

**Comment by rshriram** [comment]
Alternatively, we could have a dumb periodic check logic as an alternative - check for config changes every 2s and push them. Users can choose whichever one they prefer. 

Or an even crazier one - make people use istioctl to push the large config. Istioctl can read the entire file and count num configs in the file. It can then stick couple of annotations in each config 
 ConfigsFromFile: blah.yaml
 NumConfigsInFile: 100

We can look for these annotations and literally wait for about 80 config events from same file before pushing

**Comment by howardjohn** [comment]
Its not just `kubectl apply LARGE_FILE` but also something like `kubectl scale deployment --replicas=LARGE_NUMBER` or something similar. Although I guess that matters less since most of the load is in bootstrapping those new replicas which we don't debounce.

The config annotation may be too special case since it could be a controller spinning up a bunch of resources, helm creating them, etc. Lot of possibilities.

It is a hard problem to solve since we don't want to go too aggressive on debouncing or config is super slow to propogate. We can also consider separating EDS debounce (there is a flag to turn off EDS debounce currently)

**Comment by ramaraochavali** [comment]
I agree this is hard and we need test it out and see what works best.  Let us wait till that perf thing comes and we can tweak this as needed.

> We can also consider separating EDS debounce (there is a flag to turn off EDS debounce currently)

Are you suggesting eds follow a separate debouncer from main debouncer? If yes, that seems like a good thing to try?

**Comment by mandarjog** [comment]
Do we have a target for config propagation?
Generally I don’t see a need for propagation to be very fast. Order of several seconds seems fine.

I agree EDS should be separated. Perhaps not related to this, but we should prioritize Eds updates over other types of updates. This came up during “what is the fault behaviour when pilot is overloaded.” I think we want EDS to take precedence.

Do we already treat 0 to 1 endpoint transition different than say 2 to 3?

**Comment by ramaraochavali** [comment]
> Do we already treat 0 to 1 endpoint transition different than say 2 to 3?

No. We do not

**Comment by ramaraochavali** [comment]
Closing this and we can work when perf test is available
